viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -333,97 +371,82 @@ void SLAMServiceImpl::GetLatestSampledPointCloudMapString(
+                continue;
+            }
 
-    // Sample the image based on the number of pixels. Output is the number of
-    // pixels to skip. skip_count will reduce the size of the PCD to under 32
-    // MB, with additional tuning provided by the samplingFactor. If the PCD
-    //  would already be smaller than 32MB, do not sample the image.
-    // When moving to streaming this behavior may change.
-    int skip_count = (size_data * pixelBytetoPCDByte) / maximumGRPCByteLimit *
-                     samplingFactor;
-    if (skip_count == 0) {
-        skip_count = 1;
-    }
-
-    std::string data_buffer;
-
-    // Loop to sample data and reduce resolution. Increments multiplied
-    // by 4 to represent 4 bytes per pixel
-    for (int i = 0; i < size_data; i += skip_count * 4) {
-        // skip pixels that are not in our map(black/past walls)
-        // this check represents [102,102,102]
-        if ((data_vect[i + 0] == defaultCairosEmptyPaintedSlice) &&
-            (data_vect[i + 1] == defaultCairosEmptyPaintedSlice) &&
-            (data_vect[i + 2] == defaultCairosEmptyPaintedSlice))
-            continue;
-
-        // Determine probability based on color pixel
-        int prob = viam::ViamColorToProbability((int)data_vect[i + 2]);
-        if (prob == 0) continue;
-
-        num_points++;
-
-        int pixel_index = i / 4;
-        int pixel_x = pixel_index % width;
-        int pixel_y = pixel_index / width;
-
-        // Convert pixel location to pointcloud point in meters
-        float x_pos = (pixel_x - painted_slices->origin.x()) * kPixelSize;
-        // Y is inverted to match output from getPosition()
-        float y_pos = -(pixel_y - painted_slices->origin.y()) * kPixelSize;
-        // 2D SLAM so Z is set to 0
-        float z_pos = 0;
-
-        // Turn the map point into a vector to perform transformations with.
-        // Current transformation rotates coordinates to match slam service
-        // expectation (XZ plane)
-        Eigen::Vector3d map_point(x_pos, y_pos, z_pos);
-        auto rotated_map_point = pcdRotation * map_point;
+            // Determine probability based on the color of the pixel and skip if
+            // it is 0
+            int prob = CalculateProbabilityFromColorChannels(pixel_color);
+            if (prob == 0) {
+                continue;
+            }
 
-        viam::utils::writeFloatToBufferInBytes(data_buffer,
-                                               rotated_map_point.x());
-        viam::utils::writeFloatToBufferInBytes(data_buffer,
-                                               rotated_map_point.y());
-        viam::utils::writeFloatToBufferInBytes(data_buffer,
-                                               rotated_map_point.z());
-        viam::utils::writeIntToBufferInBytes(data_buffer, prob);
+            // Convert pixel location to pointcloud point in meters
+            float x_pos = (pixel_x - origin_pixel_x) * resolutionMeters;
```

Comment:
[q] One more (maybe silly question) but I don't understand how this reduces the resolution of the map.

I would expect that `reducing the resolution` involves representing multiple points within the resolution (5cm) as a single point with some aggregate of the probabilities of all points that were within that 2d area. 

It is not clear to me how multiplying every point with a probability by `0.05` does that.

1. Aren't we still returning the same number of points?
2. Where are we aggregating the probabilities of multiple points within the same resolution area?

Reply:
We synced offline:
The answer is that `viam::io::PaintSubmapSlices` takes the resolution & calls cairo with it to scale the painted map. Cairo then does points 1 & 2. We need to multiply by resolution on this line to convert from pixels back into meters.

Reply:
Resolved offline. The resolution comes into effect during the painting. This use of it here is jsut for converting back to realworld space.

X (meters) = Y (pixels) * resolution (meter/pixel)